### ISSUE TITLE
Prompt for missing variables in "terraform plan"

### ISSUE THREAD

**Issue by radeksimko** [hashicorp/terraform#1899]
If there are no `default` values for defined variables, terraform now ends up with an error, e.g.

```
$ terraform plan
```
```
There are warnings and/or errors related to your configuration. Please
fix these before continuing.

Errors:

  * Required variable not set: key_name
  * Required variable not set: key_path
```

I'd expect Terraform to use the same prompt as `apply` uses (with the same option `-input=false` to keep the current behaviour). Is there any particular reason why `plan` works differently here?

Related: https://github.com/hashicorp/terraform/pull/1896

**Comment by mitchellh** [comment]
Sounds good, note for future labeling: please always try to label with two labels: [bug/enhancement] and [core or provider/foo, etc.]. I triage very heavily with two labels. :)

**Comment by radeksimko** [comment]
> please always try to label with two labels

I usually try, but sometimes I'm not sure if it's a bug or a feature (enhancement) :smiley: 
I will take my best guess next time, instead of leaving 1 label then.

**Comment by askhogan** [comment]
+1 this is a helpful add.  I spent about 30 minutes this weekend digging, because of confusing output in terraform plan, which is recommended by the getting started example docs.

**Comment by pikeas** [comment]
+1, docs state that `terraform plan` will prompt for missing variables, which does not appear to be the case with 0.5.3 or current master.

**Comment by pikeas** [comment]
There's an additional wrinkle regarding environment variables. If I have:

```
provider "aws" {
    access_key = "${var.access_key}"
    secret_key = "${var.secret_key}"
    region = "us-east-1"
}

variable "access_key" {
    default = "foo"
}
variable "secret_key" {
    default = "foo"
}
```

My expectation is that when environment variables are passed (`AWS_ACCESSKEY=bar terraform plan`), those variables won't be prompted. Either that, or the prompt is retained, but the value from the env var overrides the default.

This might be a separate issue from the above. Let me know if I should open a new one.

**Comment by phobos182** [comment]
+1 Our preferred way to use terraform is to run a 'plan', and output to an execution file. Then `terraform apply` that execution plan. Not prompting for variables throws a wrinkle in that workflow.

**Comment by radeksimko** [comment]
This is already fixed in `master` (verified by running an example code from https://github.com/hashicorp/terraform/issues/1230 ).

**Comment by jmehnle** [comment]
@radeksimko, in my understanding this (#1899) is about `terraform plan` never prompting for required variables, whereas #1230 is about `terraform` (not just `terraform plan`) not prompting for required variables *if* a `tfvars` file is read. Example code passing for #1230 doesn't prove #1899 is fixed. As far as I can tell #1899 is *not* fixed in Terraform 0.6.14.
